File: backend/Infrastructure/Hubs/ArenaWebRtcService.cs
using System.Collections.Concurrent;
using System.Text.Json;
using Application.EventBus;
using Application.EventBus.Events;
using Application.Factories.Game;
using Application.Services.Game;
using Microsoft.AspNetCore.SignalR;
using Microsoft.Extensions.Logging;
using SIPSorcery.Net;

namespace Infrastructure.Hubs;

public class ArenaWebRtcService
{
    private readonly ConcurrentDictionary<string, RTCPeerConnection> _peerConnections = new();
    private readonly ConcurrentDictionary<string, RTCDataChannel> _dataChannels = new();
    private readonly ConcurrentDictionary<string, List<RTCIceCandidateInit>> _iceBuffer = new();
    
    private readonly IArenaService _arenaService;
    private readonly IArenaMessageHandlerFactory _handlerFactory;
    private readonly ILogger<ArenaWebRtcService> _logger;

    public ArenaWebRtcService(
        IArenaService arenaService,
        IArenaMessageHandlerFactory handlerFactory,
        ILogger<ArenaWebRtcService> logger,
        IEventBus eventBus)
    {
        _arenaService = arenaService;
        _handlerFactory = handlerFactory;
        _logger = logger;

        // Подписка на событие SnapshotReady через EventBus
        eventBus.Subscribe<SnapshotReadyEvent>(async e =>
        {
            if (_dataChannels.TryGetValue(e.ClientId, out var dc) &&
                dc.readyState == RTCDataChannelState.open)
            {
                var options = new JsonSerializerOptions
                {
                    PropertyNamingPolicy = JsonNamingPolicy.CamelCase
                };
                
                var packet = new { type = "snapshotResponse", payload = e.State };
                var json = JsonSerializer.Serialize(packet, options);
                dc.send(json);
            }
        });
    }

    // -------------------------------------------
    // Handle WebRTC Offer from client
    // -------------------------------------------
    public async Task<string> HandleOfferFromClientAsync(
        string clientId,
        string sdpOffer,
        IClientProxy clientProxy)
    {
        _logger.LogInformation($"[{nameof(ArenaWebRtcService)}]: Offer received from {clientId}");

        var pcConfig = new RTCConfiguration
        {
            iceServers = new[] { new RTCIceServer { urls = "stun:stun.l.google.com:19302" } }.ToList()
        };

        var pc = new RTCPeerConnection(pcConfig);
        _peerConnections[clientId] = pc; // Регистрируем сразу (до await)

        // --- Подписки на события PeerConnection ---
        pc.onicecandidate += candidate =>
        {
            if (candidate is not null)
            {
                var candidateJson = JsonSerializer.Serialize(candidate);
                clientProxy.SendAsync("ServerIceCandidate", candidateJson);
            }
        };

        pc.ondatachannel += dc =>
        {
            _logger.LogInformation($"[{nameof(ArenaWebRtcService)}]: DataChannel '{dc.label}' created for {clientId}");
            _dataChannels[clientId] = dc;

            dc.onopen += async () =>
            {
                _logger.LogInformation($"[{nameof(ArenaWebRtcService)}]: DataChannel '{dc.label}' opened for {clientId}");
                await SendSnapshotToClientAsync(clientId);
            };

            dc.onmessage += async (_, _, data) =>
            {
                var json = System.Text.Encoding.UTF8.GetString(data);
                await HandleClientMessageAsync(clientId, json);
            };

            dc.onclose += () =>
            {
                _logger.LogInformation($"[{nameof(ArenaWebRtcService)}]: DataChannel '{dc.label}' closed for {clientId}");
                _dataChannels.TryRemove(clientId, out _);
            };
        };

        // --- SDP Setup ---
        var remoteDesc = new RTCSessionDescriptionInit
        {
            type = RTCSdpType.offer,
            sdp = sdpOffer
        };

        pc.setRemoteDescription(remoteDesc);

        var answer = pc.createAnswer();
        await pc.setLocalDescription(answer);

        // --- Применяем отложенные ICE кандидаты ---
        if (_iceBuffer.TryRemove(clientId, out var buffered))
        {
            foreach (var c in buffered)
                pc.addIceCandidate(c);
        }

        _logger.LogInformation($"[{nameof(ArenaWebRtcService)}]: Answer created for {clientId}");
        return answer.sdp;
    }

    // -------------------------------------------
    // ICE Candidate from client
    // -------------------------------------------
    public void AddIceCandidate(string clientId, string candidateJson)
    {
        var candidate = JsonSerializer.Deserialize<RTCIceCandidateInit>(candidateJson);
        if (candidate is null) return;

        if (_peerConnections.TryGetValue(clientId, out var pc))
        {
            pc.addIceCandidate(candidate);
        }
        else
        {
            var buffer = _iceBuffer.GetOrAdd(clientId, _ => new List<RTCIceCandidateInit>());
            buffer.Add(candidate);
            _logger.LogDebug($"[{nameof(ArenaWebRtcService)}]: Buffered ICE candidate for {clientId}");
        }
    }

    // -------------------------------------------
    // Graceful Close
    // -------------------------------------------
    public void ClosePeer(string clientId)
    {
        if (_dataChannels.TryRemove(clientId, out var dc))
        {
            try { dc.close(); }
            catch
            {
                // ignored
            }
        }

        if (_peerConnections.TryRemove(clientId, out var pc))
        {
            try { pc.close(); }
            catch
            {
                // ignored
            }
        }

        _logger.LogInformation($"[{nameof(ArenaWebRtcService)}]: Closed peer {clientId}");
    }

    // -------------------------------------------
    // Handle messages from client
    // -------------------------------------------
    private async Task HandleClientMessageAsync(string clientId, string message)
    {
        var doc = JsonDocument.Parse(message);

        if (!doc.RootElement.TryGetProperty("type", out var t)) return;
        var type = t.GetString() ?? string.Empty;
        var payload = doc.RootElement.TryGetProperty("payload", out var p) ? p : default;

        var handler = _handlerFactory.CreateHandler(type);
        if (handler is null)
        {
            _logger.LogWarning($"[{nameof(ArenaWebRtcService)}]: No handler for message type '{type}'");
            return;
        }

        await handler.HandleAsync(clientId, payload);
    }

    // -------------------------------------------
    // Send snapshot immediately via data channel
    // -------------------------------------------
    private async Task SendSnapshotToClientAsync(string clientId)
    {
        if (!_dataChannels.TryGetValue(clientId, out var dc)) return;
        if (dc.readyState != RTCDataChannelState.open) return;

        var state = await _arenaService.GetStateAsync(clientId);
        var packet = new { type = "snapshotResponse", payload = state };
        var json = JsonSerializer.Serialize(packet);
        dc.send(json);

        _logger.LogDebug($"[{nameof(ArenaWebRtcService)}]: Snapshot sent to {clientId}");
    }
}

File: client/src/ws/arenaWebRTCService.ts
import * as signalR from "@microsoft/signalr";
import type { ArenaState } from "../@types/game/arenaState";

// ---------------------------
// Types
// ---------------------------

// Payload for "move" messages
type MovePayload = { x: number; y: number };

// Payload for "snapshotResponse" messages
type SnapshotResponsePayload = ArenaState;

// Union of all possible messages
type MessagePayload =
    | { type: "move"; payload: MovePayload }
    | { type: "getSnapshot" }
    | { type: "snapshotResponse"; payload: SnapshotResponsePayload };

// Messages that the client can send
type ClientMessage = Extract<MessagePayload, { type: "move" | "getSnapshot" }>;

// ---------------------------
// ArenaService
// ---------------------------
class ArenaService {
    private static instance: ArenaService | null = null;

    /** SignalR connection for WebRTC signaling */
    private connection: signalR.HubConnection | null = null;

    /** WebRTC peer connection */
    private peerConnection: RTCPeerConnection | null = null;

    /** DataChannel for sending/receiving game messages */
    private dataChannel: RTCDataChannel | null = null;

    /** Listeners for arena state updates */
    private listeners: ((state: ArenaState) => void)[] = [];

    /** Private constructor for singleton pattern */
    private constructor() {}

    /** Get singleton instance of ArenaService */
    public static getInstance(): ArenaService {
        if (!ArenaService.instance) {
            ArenaService.instance = new ArenaService();
        }
        return ArenaService.instance;
    }

    // ---------------------------
    // Connect to SignalR and initialize WebRTC
    // ---------------------------
    public async connect() {
        if (this.connection) return;

        // Initialize SignalR connection for signaling
        this.connection = new signalR.HubConnectionBuilder()
            .withUrl("http://localhost:5184/arenaSignalHub")
            .withAutomaticReconnect()
            .build();

        /** Handle SDP answer from server */
        this.connection.on("AnswerFromServer", async (sdpAnswer: string) => {
            if (!this.peerConnection) return;

            const answer: RTCSessionDescriptionInit = {
                type: "answer" as RTCSdpType,
                sdp: sdpAnswer,
            };

            await this.peerConnection.setRemoteDescription(answer);
        });

        /** Handle ICE candidates from server */
        this.connection.on("ServerIceCandidate", async (candidateJson: string) => {
            if (!this.peerConnection) return;
            const candidate = JSON.parse(candidateJson) as RTCIceCandidateInit;
            await this.peerConnection.addIceCandidate(candidate);
        });

        await this.connection.start();
        console.log("Connected to ArenaSignalHub");

        await this.initWebRtc();
    }

    // ---------------------------
    // Initialize WebRTC PeerConnection and DataChannel
    // ---------------------------
    private async initWebRtc() {
        this.peerConnection = new RTCPeerConnection({
            iceServers: [{ urls: "stun:stun.l.google.com:19302" }],
        });

        // Create DataChannel for game messages
        this.dataChannel = this.peerConnection.createDataChannel("arena");
        this.dataChannel.onopen = () => console.log("DataChannel opened");
        this.dataChannel.onclose = () => console.log("DataChannel closed");

        /** Handle incoming messages from server */
        this.dataChannel.onmessage = (event) => {
            const msg = JSON.parse(event.data) as MessagePayload;

            if (msg.type === "snapshotResponse") {
                this.listeners.forEach((fn) => fn(msg.payload));
            }
        };

        /** Handle ICE candidates locally and send to server */
        this.peerConnection.onicecandidate = (event) => {
            if (event.candidate) {
                this.connection?.invoke(
                    "SendIceCandidateToServer",
                    JSON.stringify(event.candidate)
                );
            }
        };

        // Create SDP offer
        const offer = await this.peerConnection.createOffer();
        await this.peerConnection.setLocalDescription(offer);

        // Send SDP offer to server via SignalR
        await this.connection?.invoke("OfferToServer", offer.sdp!);
    }

    // ---------------------------
    // Send a message to server via DataChannel
    // ---------------------------
    public sendMessage(message: ClientMessage) {
        if (!this.dataChannel || this.dataChannel.readyState !== "open") return;
        this.dataChannel.send(JSON.stringify(message));
    }

    /** Move player to a new position */
    public moveTo(x: number, y: number) {
        this.sendMessage({ type: "move", payload: { x, y } });
    }

    /** Request arena snapshot from server */
    public getSnapshot() {
        this.sendMessage({ type: "getSnapshot" });
    }

    /** Subscribe to arena state updates */
    public onUpdate(callback: (state: ArenaState) => void) {
        this.listeners.push(callback);
    }
}

// ---------------------------
// Export singleton instance
// ---------------------------
export const arenaWebRTCService = ArenaService.getInstance();